File: packages/agent-python/bolthub/_invoice.py
"""Dependency-free BOLT11 amount decoding.

Only the human-readable prefix (HRP) of a BOLT11 invoice is parsed to recover
the payable amount. The bech32 data part (which carries payment hash, routing
hints, etc.) is never decoded here, so no external dependency is required.

This is the authoritative price source for an L402 challenge: the invoice is
always present in the ``WWW-Authenticate`` header, and the bolthub gateway's own
end-to-end test asserts the invoice amount equals the body's ``amountSats``
(see ``packages/cli/e2e/paid-call.ts``).

Note on correctness: a naive ``^ln(bc|tb...)(\\d+)([munp])`` regex mis-reads an
*amountless* invoice (``lnbc1p...``) by treating the bech32 separator ``1`` as
an amount digit. We instead split the HRP off at the bech32 separator (the last
``1`` — the bech32 data charset never contains ``1``) and parse the amount from
the HRP alone, which disambiguates amountless invoices to ``None``.
"""

from __future__ import annotations

import re
from decimal import Decimal, ROUND_HALF_UP

# Currency prefixes, longest-first so the alternation is unambiguous:
# bcrt (regtest), bc (mainnet), tbs (signet), tb (testnet).
_HRP_RE = re.compile(r"^ln(?:bcrt|bc|tbs|tb)(\d+)([munp])$")

# sats = digits * multiplier_in_btc * 1e8.
#   m (milli, 1e-3) -> * 1e5
#   u (micro, 1e-6) -> * 1e2
#   n (nano,  1e-9) -> / 10
#   p (pico,  1e-12)-> / 10_000
_SATS_PER_UNIT = {
    "m": Decimal(100_000),
    "u": Decimal(100),
    "n": Decimal(1) / Decimal(10),
    "p": Decimal(1) / Decimal(10_000),
}


def bolt11_amount_sats(invoice: str) -> int | None:
    """Decode the amount in satoshis from a BOLT11 invoice's HRP.

    Returns ``None`` for amountless invoices, multiplier-less (whole-BTC)
    amounts, or anything that does not parse. Sub-satoshi amounts (``p``
    multiplier) are rounded to the nearest satoshi.
    """
    if not invoice:
        return None
    s = invoice.strip().lower()
    # The bech32 separator is the last '1'; bech32 data never contains '1'.
    if "1" not in s:
        return None
    hrp = s.rsplit("1", 1)[0]
    m = _HRP_RE.match(hrp)
    if not m:
        return None
    digits, unit = m.group(1), m.group(2)
    sats = Decimal(digits) * _SATS_PER_UNIT[unit]
    value = int(sats.to_integral_value(rounding=ROUND_HALF_UP))
    return value if value > 0 else None

File: packages/pay/src/http/invoice.ts
/**
 * Dependency-free BOLT11 amount decoding.
 *
 * Only the human-readable prefix (HRP) of a BOLT11 invoice is parsed to recover
 * the payable amount. The bech32 data part is never decoded, so no dependency
 * is required. The invoice is the authoritative price source for an L402
 * challenge: it is always present in the `WWW-Authenticate` header, and the
 * gateway's own e2e asserts the invoice amount equals the body's `amountSats`.
 *
 * The HRP is split off at the bech32 separator (the last `1`; bech32 data never
 * contains `1`) so an amountless invoice (`lnbc1p...`) resolves to `null` rather
 * than mis-reading the separator as an amount.
 */

// Currency prefixes, longest-first: bcrt (regtest), bc (mainnet), tbs (signet), tb (testnet).
const HRP_RE = /^ln(?:bcrt|bc|tbs|tb)(\d+)([munp])$/;

// sats = digits * multiplier_in_btc * 1e8.
const SATS_PER_UNIT: Record<string, number> = {
  m: 100_000, // milli (1e-3)
  u: 100, // micro (1e-6)
  n: 0.1, // nano  (1e-9)
  p: 0.0001, // pico  (1e-12)
};

/**
 * Decode the amount in satoshis from a BOLT11 invoice's HRP. Returns `null` for
 * amountless invoices, multiplier-less amounts, or anything that does not parse.
 * Sub-satoshi amounts are rounded to the nearest satoshi.
 */
export function bolt11AmountSats(invoice: string): number | null {
  if (!invoice) return null;
  const s = invoice.trim().toLowerCase();
  // The bech32 separator is the last `1`; the HRP is everything before it.
  const sep = s.lastIndexOf("1");
  if (sep <= 0) return null;
  const hrp = s.slice(0, sep);
  const m = HRP_RE.exec(hrp);
  if (!m) return null;
  const sats = Math.round(Number(m[1]) * SATS_PER_UNIT[m[2]]);
  return sats > 0 ? sats : null;
}
